### PR TITLE
Fix/memory usage

### PR DESCRIPTION
**Pull request by TrevorBurgoyne**
# Fix/memory usage
[WS-914](https://sentera.atlassian.net/browse/WS-914)
## Description
Recently I have found that ULabel's performance rapidly deteriorates with large numbers of annotations and large images. Memory usage would spike and cause crashes, and all operations got very laggy. I tackled this in three main areas:
- First, in #133 I worked on decreasing the number of redraws necessary by separating each annotation onto its own canvas context. This was fine for a small number of annotations, however stacking many canvases became the new root cause of many of the memory and performance issues when  working with a large number of annotations. 
  - Now instead on one canvas per annotation, I've added a `N_ANNOS_PER_CANVAS` constant set to 100, and now groups of 100 annotations are drawn on a single canvas. For large numbers of annotations, this retains some of the improvements of #133 by only requiring a redraw of (at max) 100 annotations per operation that targets a single annotation, while drastically cutting the number of canvases required, which significantly helps performance.
- Second, tried to cut down on the number of "deep copies" (`JSON.parse(JSON.stringify())`) where possible to reduce superfluous memory usage   
- And lastly, a slight rework of `undo`/`redo` payloads, using `JSON.stringify()` when recording actions and waiting to parse them until needed. Also fixed `undo`/`redo` behavior of polygons, closing #146   

On the big dataset I was testing on, `v0.10.11` was using up to 6000 MB of memory and was noticeably laggy
Now with `v0.10.12`, only 1000 MB memory and almost no lag 

## PR Checklist

- [x] Merged latest main
- [x] Version number in `package.json` has been bumped since last release
- [x] Version numbers match between package `package.json` and `src/version.js`
- [x] Ran `npm install` and `npm run build` AFTER bumping the version number
- [x] Updated documentation if necessary (currently just in `api_spec.md`)
- [x] Added changes to `changelog.md`

## Breaking API Changes
none

[WS-914]: https://sentera.atlassian.net/browse/WS-914?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1689,13 +1690,28 @@ export class ULabel {
         }
     }
 
-    // Create a new canvas for an individual annotation and return its context
-    get_init_canvas_context_id(annotation_id, subtask = null) {
+    // Find the next available annotation context and return its ID
+    // If all contexts are in use, create a new one
+    get_next_available_canvas_id(subtask = null) {
```
Comment (csolbs24) [inline src/index.js:1695]:
```suggestion
    /**
     * Find the next available annotation context and return its ID.
     * If all annotation contexts are in use, create a new canvas and return it's id.
     * @param {ULabelSubtask} subtask 
     * @returns {string} The ID of an available canvas.
     */
    get_next_available_canvas_id(subtask = null) {
```
I know that we haven't been very consistent with this, but we should move towards giving every function/method a docstring with a brief overview of its purpose. [Here](https://google.github.io/styleguide/jsguide.html#jsdoc) is documentation outlining JavaScript best practices as a whole, but section 7 is the section on docstrings.

File: src/index.js [hunk 2]
```diff
@@ -1689,13 +1690,28 @@ export class ULabel {
...
+            if (this.subtasks[subtask]["state"]["annotation_contexts"][canvas_ids[i]]["annotation_ids"].length < N_ANNOS_PER_CANVAS) {
+                return canvas_ids[i];
+            }
+        }
+        // If no canvas has less than N_ANNOS_PER_CANVAS annotations, create a new canvas
+        return this.create_annotation_canvas(subtask);
+    }
+
+    // Create a new canvas and return its ID
+    create_annotation_canvas(subtask) {
```
Comment (csolbs24) [inline src/index.js:1711]:
```suggestion
    /**
     * Create a new canvas for annotations and return its ID
     * @param {ULabelSubtask } subtask 
     * @returns {string} The ID of the new canvas
     */
    create_annotation_canvas(subtask) {
```
Another example.

File: src/index.js [hunk 3]
```diff
@@ -1710,7 +1726,23 @@ export class ULabel {
...
+        this.subtasks[subtask]["state"]["annotation_contexts"][canvas_id] = {
+            "annotation_ids": [],
+            "context": document.getElementById(canvas_id).getContext("2d")
+        };
+
+        return canvas_id;
+    }
+
+    // Create a new canvas for an individual annotation and return its context
+    get_init_canvas_context_id(annotation_id, subtask = null) {
```
Comment (csolbs24) [inline src/index.js:1738]: Docstring

File: src/index.js (hunk 1 above)
Comment (TrevorBurgoyne) [reply src/index.js]: [9d64b2d](https://github.com/SenteraLLC/ulabel/pull/152/commits/9d64b2d52e57d44b18106b2f7784804d691cfc91)

File: src/index.js (hunk 2 above)
Comment (TrevorBurgoyne) [reply src/index.js]: [9d64b2d](https://github.com/SenteraLLC/ulabel/pull/152/commits/9d64b2d52e57d44b18106b2f7784804d691cfc91)

File: src/index.js (hunk 3 above)
Comment (TrevorBurgoyne) [reply src/index.js]: [9d64b2d](https://github.com/SenteraLLC/ulabel/pull/152/commits/9d64b2d52e57d44b18106b2f7784804d691cfc91)

File: src/annotation.ts [hunk 4]
```diff
@@ -12,6 +12,14 @@ export const DELETE_MODES = ["delete_polygon", "delete_bbox"]
 export const DELETE_CLASS_ID = -1;
 export const MODES_3D = ["global", "bbox3"];
 export const NONSPATIAL_MODES = ["whole-image", "global"];
+export const N_ANNOS_PER_CANVAS = 20;
```
[the remark this answers is not in the document]
Comment (TrevorBurgoyne) [reply src/annotation.ts:15]: tbh it was kinda arbitrary, I tried it first bc 10 seemed too low. Ideally it should be as high as possible such that there's no noticeable lag when working with < `N_ANNOS_PER_CANVAS` annotations, I'll test some higher values and see what works

[the remark this answers is not in the document]
Comment (TrevorBurgoyne) [reply src/annotation.ts]: k i did a better param sweep from 10 -> 200, and the sweet spot was 100 on an image with a bunch of polygon annotations. Higher than that and I started to notice some lag when drawing

[89663f4](https://github.com/SenteraLLC/ulabel/pull/152/commits/89663f4820a3aa588b5857529d7e141952708a63)

**Pull request merged**